File: tests/Tests.CSharp/FastCdcCrossVerifyTests.cs
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp;

namespace Zeta.Tests.CSharp;

/// <summary>
/// FastCDC cross-language agreement — the C# oracle replays the shared seed
/// (<c>src/Core.TypeScript/fastcdc/golden-vectors.json</c>) that the F#/TS/Rust oracles also verify.
/// Each oracle regenerates the byte stream deterministically (byte[i] = mix(i) &amp; 0xFF) and the chunk
/// lengths are cross-verified; the large stream exercises genuine content-defined cuts.
/// </summary>
public class FastCdcCrossVerifyTests
{
    private static string RepoRoot()
    {
        var dir = new DirectoryInfo(Path.GetDirectoryName(typeof(FastCdcCrossVerifyTests).Assembly.Location)!);
        while (dir is not null && !File.Exists(Path.Join(dir.FullName, "Zeta.sln")))
        {
            dir = dir.Parent;
        }

        return dir?.FullName ?? throw new DirectoryNotFoundException("Could not locate repo root (Zeta.sln).");
    }

    private static JsonElement Seed()
    {
        var path = Path.Join(RepoRoot(), "src", "Core.TypeScript", "fastcdc", "golden-vectors.json");
        Assert.True(File.Exists(path), $"seed not found: {path}");
        using var doc = JsonDocument.Parse(File.ReadAllText(path));
        return doc.RootElement.Clone();
    }

    [Fact]
    public void GearSamplesAgreeWithSeed()
    {
        var table = FastCdc.GearTable();
        foreach (var v in Seed().GetProperty("gearSamples").EnumerateArray())
        {
            Assert.Equal(ulong.Parse(v.GetProperty("value").GetString()!, CultureInfo.InvariantCulture),
                table[v.GetProperty("i").GetInt32()]);
        }
    }

    [Fact]
    public void ChunkLengthsAgreeWithSeed()
    {
        foreach (var v in Seed().GetProperty("chunk").EnumerateArray())
        {
            var expected = v.GetProperty("lengths").EnumerateArray().Select(e => e.GetInt32()).ToArray();
            var bytes = FastCdc.GenBytes(v.GetProperty("count").GetInt32());
            var got = FastCdc.ChunkLengths(
                bytes,
                v.GetProperty("min").GetInt32(),
                v.GetProperty("avg").GetInt32(),
                v.GetProperty("max").GetInt32());
            Assert.Equal<IReadOnlyList<int>>(expected, got);
        }
    }
}

File: tests/Tests.FSharp/FastCdc.CrossVerify.Tests.fs
module Zeta.Tests.FastCdcCrossVerifyTests

open System
open System.IO
open System.Text.Json
open global.Xunit
open Zeta.Core

// ═══════════════════════════════════════════════════════════════════
// FastCDC cross-language agreement — the F# oracle uses the REAL canonical FastCdcChunker and replays
// the shared seed (src/Core.TypeScript/fastcdc/golden-vectors.json) that the C#/TS/Rust oracles also
// verify. Each oracle regenerates the byte stream deterministically (byte[i] = SplitMix64.mix(i) & 0xFF)
// and the chunk LENGTHS are cross-verified; the 200000-byte stream exercises genuine content-defined
// cuts (variable lengths, not just max-forced). The private Gear table is locked directly by the
// C#/TS/Rust oracles and transitively here (matching lengths imply a matching table).
// ═══════════════════════════════════════════════════════════════════

let private repoRoot () =
    let mutable dir = DirectoryInfo(Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location))
    while not (isNull dir) && not (File.Exists(Path.Join(dir.FullName, "Zeta.sln"))) do
        dir <- dir.Parent
    if isNull dir then failwith "Could not locate repo root (Zeta.sln)." else dir.FullName

let private genBytes (count: int) : byte[] =
    Array.init count (fun i -> byte (SplitMix64.mix (uint64 i) &&& 0xFFUL))

[<Fact>]
let ``F# FastCdcChunker agrees with the shared golden seed`` () =
    let path = Path.Join(repoRoot (), "src", "Core.TypeScript", "fastcdc", "golden-vectors.json")
    Assert.True(File.Exists path, sprintf "seed not found: %s" path)
    use doc = JsonDocument.Parse(File.ReadAllText path)
    let chunks = doc.RootElement.GetProperty("chunk").EnumerateArray() |> Seq.toArray

    for v in chunks do
        let count = (v.GetProperty "count").GetInt32()
        let min = (v.GetProperty "min").GetInt32()
        let avg = (v.GetProperty "avg").GetInt32()
        let max = (v.GetProperty "max").GetInt32()
        let expected = [| for l in (v.GetProperty "lengths").EnumerateArray() -> l.GetInt32() |]

        let chunker = FastCdcChunker(min, avg, max)
        chunker.Push(ReadOnlySpan<byte>(genBytes count))
        chunker.Flush()
        let got = chunker.DrainChunks() |> Array.map (fun c -> c.Length)
        Assert.Equal<int[]>(expected, got)
